controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentUpgraderTest.java

```diff
@@ -0,0 +1,101 @@
+        Version systemVersion = tester.controller().readSystemVersion();
+        Instant start = tester.clock().instant().truncatedTo(MILLIS);
+
+        devApp.runJob(devUsEast1, appPackage);
+        prodApp.submit(appPackage).deploy();
+        assertEquals(systemVersion, tester.jobs().last(devApp.instanceId(), devUsEast1).get().versions().targetPlatform());
+        assertEquals(systemVersion, tester.jobs().last(prodApp.instanceId(), productionUsWest1).get().versions().targetPlatform());
+
+        // Not upgraded initially
+        upgrader.maintain();
+        assertEquals(start, tester.jobs().last(devApp.instanceId(), devUsEast1).get().start());
+        assertEquals(start, tester.jobs().last(prodApp.instanceId(), productionUsWest1).get().start());
+
+        // Not upgraded immediately after system upgrades
+        tester.controllerTester().upgradeSystem(new Version(7, 8, 9));
+        upgrader.maintain();
+        assertEquals(start, tester.jobs().last(devApp.instanceId(), devUsEast1).get().start());
+        assertEquals(start, tester.jobs().last(prodApp.instanceId(), productionUsWest1).get().start());
+
+        // 13 hours pass, but not upgraded before a day has passed since last deployment
+        tester.clock().advance(Duration.ofHours(13));
+        upgrader.maintain();
+        assertEquals(start, tester.jobs().last(devApp.instanceId(), devUsEast1).get().start());
+        assertEquals(start, tester.jobs().last(prodApp.instanceId(), productionUsWest1).get().start());
+
+        // 34 hours pass, but not upgraded since it's not likely in the middle of the night
+        tester.clock().advance(Duration.ofHours(21));
+        upgrader.maintain();
+        assertEquals(start, tester.jobs().last(devApp.instanceId(), devUsEast1).get().start());
+        assertEquals(start, tester.jobs().last(prodApp.instanceId(), productionUsWest1).get().start());
+
+        // 37 hours pass, and the dev deployment, only, is upgraded
+        tester.clock().advance(Duration.ofHours(3));
+        upgrader.maintain();
+        assertEquals(tester.clock().instant().truncatedTo(MILLIS), tester.jobs().last(devApp.instanceId(), devUsEast1).get().start());
+        assertTrue(tester.jobs().last(devApp.instanceId(), devUsEast1).get().isRedeployment());
+        assertEquals(start, tester.jobs().last(prodApp.instanceId(), productionUsWest1).get().start());
+        devApp.runJob(devUsEast1);
+
+        // After the upgrade, the dev app is mostly (re)deployed to at night, but this doesn't affect what is likely the night.
+        tester.controllerTester().upgradeSystem(new Version(7, 9, 11));
+        tester.clock().advance(Duration.ofHours(48));
+        upgrader.maintain();
+        assertEquals(tester.clock().instant().truncatedTo(MILLIS), tester.jobs().last(devApp.instanceId(), devUsEast1).get().start());
+    }
+
+    @Test
+    public void testNight() {
+        assertEquals(14, mostLikelyWeeHour(new int[]{ 0, 1, 2, 3, 4, 5, 6 }));
```

Comment:
Not that it's very important, but I think its easier and you get better result if you just count deployment frequencies by hour and then minimize product sum of `numDeployments(hour + offset) / (offset + 1)`, i.e.:
```
static int mostLikelyWeeHour(int[] starts) {
    int[] buckets = new int[24];
    for (int start : starts)
        buckets[start]++;

    int best = -1;
    double min = Double.MAX_VALUE;
    for (int i = 12; i < 36; i++) {
        double sum = buckets[i % 24];
        for (int j = 1; j < 13; j++)
            sum += (buckets[(i + j) % 24] + buckets[(i - j) % 24]) / (j + 1d);

        if (sum < min) {
            min = sum;
            best = i;
        }
    }
    return best % 24;
}
```
The expectation for test cases then becomes 15, 13, 15 and 18. So 0, 1, 2, ..., 6 gives the same hour as 6, 5, 4, ..., 0 and 0, 12 alternating gives hour furthest apart from them - 18.

Reply:
I like the objective function, but I'm keeping the weights :)